pkg/ebpf/perf/event.go

```diff
@@ -0,0 +1,213 @@
+	}
+	if opts.Handler == nil {
+		return nil, errors.New("invalid options: Handler is required")
+	}
+	e := &EventHandler{
+		opts: opts,
+	}
+	return e, nil
+}
+
+// Init must be called after ebpf-manager.Manager.LoadELF but before ebpf-manager.Manager.Init/InitWithOptions()
+func (e *EventHandler) Init(mgr *manager.Manager, mgrOpts *manager.Options) error {
+	ms, _, _ := mgr.GetMapSpec(e.opts.MapName)
+	if ms == nil {
+		return fmt.Errorf("unable to find map spec %q", e.opts.MapName)
+	}
+
+	if e.opts.UseRingBuffer && features.HaveMapType(ebpf.RingBuf) == nil {
+		if e.opts.UpgradePerfBuffer {
+			if ms.Type != ebpf.PerfEventArray {
+				return fmt.Errorf("map %q is not a perf buffer, got %q instead", e.opts.MapName, ms.Type.String())
+			}
+			UpgradePerfBuffer(mgr, mgrOpts, e.opts.MapName)
+		} else if ms.Type != ebpf.RingBuf {
+			return fmt.Errorf("map %q is not a ring buffer, got %q instead", e.opts.MapName, ms.Type.String())
+		}
+
+		if ms.MaxEntries != uint32(e.opts.RingBufOptions.BufferSize) {
+			ResizeRingBuffer(mgrOpts, e.opts.MapName, e.opts.RingBufOptions.BufferSize)
+		}
+		e.initRingBuffer(mgr)
+		return nil
+	}
+
+	if ms.Type != ebpf.PerfEventArray {
+		return fmt.Errorf("map %q is not a perf buffer, got %q instead", e.opts.MapName, ms.Type.String())
+	}
+	e.initPerfBuffer(mgr)
+	return nil
+}
+
+// MapType returns the ebpf.MapType of the underlying events map
+// This is only valid after calling Init.
+func (e *EventHandler) MapType() ebpf.MapType {
+	switch e.f.(type) {
+	case *manager.PerfMap:
+		return ebpf.PerfEventArray
+	case *manager.RingBuffer:
+		return ebpf.RingBuf
+	default:
+		return ebpf.UnspecifiedMap
+	}
+}
+
+// Flush flushes the pending data from the underlying perfbuf/ringbuf
+func (e *EventHandler) Flush() {
+	e.f.Flush()
+}
+
+// ResizeRingBuffer resizes the ring buffer by creating/updating a map spec editor
+func ResizeRingBuffer(mgrOpts *manager.Options, mapName string, bufferSize int) {
+	if mgrOpts.MapSpecEditors == nil {
+		mgrOpts.MapSpecEditors = make(map[string]manager.MapSpecEditor)
+	}
+	specEditor := mgrOpts.MapSpecEditors[mapName]
+	specEditor.MaxEntries = uint32(bufferSize)
+	specEditor.EditorFlag |= manager.EditMaxEntries
+	mgrOpts.MapSpecEditors[mapName] = specEditor
+}
+
+func (e *EventHandler) initPerfBuffer(mgr *manager.Manager) {
+	mgr.PerfMaps = slices.DeleteFunc(mgr.PerfMaps, func(perfMap *manager.PerfMap) bool {
+		return perfMap.Name == e.opts.MapName
+	})
+	pm := &manager.PerfMap{
+		Map: manager.Map{Name: e.opts.MapName},
+		PerfMapOptions: manager.PerfMapOptions{
+			PerfRingBufferSize: e.opts.PerfOptions.BufferSize,
+			Watermark:          e.opts.PerfOptions.Watermark,
+			WakeupEvents:       e.opts.PerfOptions.WakeupEvents,
+			RecordHandler:      e.perfRecordHandler,
+			LostHandler:        nil, // TODO do we need support for Lost?
+			RecordGetter:       perfPool.Get,
+			TelemetryEnabled:   e.opts.TelemetryEnabled,
+		},
```

Comment:
Do we have any benefit from having our own `PerfMapOptions` struct, and not simply using `manager.PerfMapOptions`?

Reply:
Yes. I don't want all the fields from manager.PerfMapOptions.